tests/network.rs

```diff
@@ -306,47 +313,153 @@ impl Network {
+
+        {
+            let node0 = self.get_node_ref(node0);
+            let node1 = self.get_node_ref(node1);
+
+            let can_connect = (node0.table.need_to_add(&node1.name) &&
+                               node1.table.allow_connection(&node0.name)) ||
+                              (node1.table.need_to_add(&node0.name) &&
+                               node0.table.allow_connection(&node1.name));
+
+            if !can_connect {
+                return actions;
             }
         }
+
+        actions.append(&mut self.connect(node0, node1));
+        actions.append(&mut self.connect(node1, node0));
+        actions
     }
 
-    fn connect_if_allowed(node0: &mut Node, node1: &mut Node) {
-        if node0.table.need_to_add(&node1.name) && node1.table.allow_connection(&node0.name) {
-            node0.connect(&node1.name, node1.endpoint);
-            node1.connect(&node0.name, node0.endpoint);
+    // Connect node0 and node1 by adding node1 to node0's routing table.
+    // This forms only half of the connection. Full connection is achieved by
+    // also calling connect(node1, node0)
+    fn connect(&mut self, node0: NodeHandle, node1: NodeHandle) -> Vec<Action> {
+        let mut actions = Vec::new();
+
+        let (node1_name, node1_endpoint) = {
+            let node1 = self.get_node_ref(node1);
+            (node1.name.clone(), node1.endpoint)
+        };
+
+        let notify_names = {
+            let node0 = self.get_node_mut_ref(node0);
+
+            if let Some(AddedNodeDetails{ must_notify, .. }) = node0.table.add(node1_name) {
+                let _ = node0.connections.insert(node1_name.clone(), Connection(node1_endpoint));
+                must_notify
+            } else {
+                Vec::new()
+            }
+        };
+
+        for notify_name in notify_names {
+            if let Some(node2) = self.find_node_by_name(&notify_name) {
+                actions.append(&mut self.connect_if_allowed(node1, node2));
+            }
+        }
+
+        actions
+    }
+
+    // Disconnect the node with `name` from `node0`.
+    fn disconnect(&mut self, node0: NodeHandle, name: &XorName) -> Vec<Action> {
+        let mut actions = Vec::new();
+
+        let incomplete_bucket = {
+            let node = self.get_node_mut_ref(node0);
+
+            if let Some(DroppedNodeDetails{ incomplete_bucket, .. }) = node.table.remove(name) {
+                let _ = node.connections.remove(&name);
+                incomplete_bucket
+            } else {
+                None
+            }
+        };
+
+        // If removing the node caused a full bucket to become not full, we have
+        // to refill it.
+        if let Some(bucket_index) = incomplete_bucket {
+            let bucket_name = match self.get_node_name(node0).with_flipped_bit(bucket_index) {
+                Ok(name) => name,
+                Err(_) => return actions,
+            };
+
+            for node1 in self.get_nodes_close_to(&bucket_name) {
```

Comment:
This `get_nodes_close_to` might return only `GROUP_SIZE - 1` nodes, if the furthest node in that bucket's close group still has `name` in its routing table. This could prevent `node0`'s bucket from being refilled.